packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -742,25 +739,106 @@ function updatePlaceholderComponent(
       }
     }
 
+    workInProgress.memoizedProps = nextProps;
+    workInProgress.memoizedState = nextDidTimeout;
+
+    // A placeholder has two sets of children: the normal children, and the
+    // fallback children that render if the normal children time out. The
+    // combined children of the placeholder are a tuple of [normalChildren,
+    // fallbackChildren]. We always render both slots, to avoid reconciliation
+    // mismatches, but a non-timed-out placeholder's fallback children are
+    // always null.
+    //
+    // When the placeholder times out, the normal children are not unmounted.
+    // We keep them in the tree to preserve their existing state. Then the
+    // renderer hides them visually e.g. using display: none.
+    //
+    // The fallback children are wrapped in an extra fiber with a special,
+    // private type. We use the type to distinguish between the two sets.
```

Comment:
@sebmarkbage This is the most important part to review. I'm not that happy with this approach; I think the ideally we would store each set of children on a separate field, the way call/return used to store its first child set on the state node. I resisted that at first because I thought it would be cumbersome to have to account for both sets every time we traverse the fiber tree (including context). But it turns out that I have to account for it in most of the places regardless.

Any alternative ideas?

Reply:
What if the stateNode was null when it is not timed out. We store normalChildren on firstChild. Then when we timeout, we swap the normalChildren and store the normalChildren on the stateNode while it is timed out and the fallbackChildren on firstChild.

Whenever we retry to see if we're unblocked we try to restore the normalChildren to firstChild first and then render down. If we don't succeed, we restore them to the placeholder again.